### PR TITLE
Plugin implementation

### PR DESCRIPTION
**Pull request by ekinhbayar**

### CODE REVIEW

File: composer.json [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+    "authors": [
+    ],
+    "minimum-stability": "dev",
+    "prefer-stable": true,
+    "require": {
+        "php": "^7.4",
+        "ext-dom": "*",
+        "amphp/amp": "^2.2.0",
+        "async-bot/core": "dev-master",
+        "room11/dom-utils": "^1.1"
```
Comment (PeeHaa) [inline composer.json]: I think this can be moved to dev requirement as it is only used in tests?

File: src/Exception/Exception.php [hunk 2]
```diff
@@ -0,0 +1,5 @@
+<?php declare(strict_types=1);
+
+namespace AsyncBot\Plugin\GoogleSearch\Exception;
+
+class Exception extends \Exception {}
```
Comment (PeeHaa) [inline src/Exception/Exception.php]: Put curly braces on their own lines

File: src/Parser/GoogleSearchParser.php [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+        if (!$mainContainer) {
+            throw new UnexpectedHtmlFormat(self::CONTAINER_ELEMENT_ID);
+        }
+
+        return new SearchResults(...$this->buildResultSet($mainContainer, $xpath));
+    }
+
+    /**
+     * @param \DOMElement $mainContainer
+     * @param \DOMXPath   $xpath
```
Comment (PeeHaa) [inline src/Parser/GoogleSearchParser.php]: Drop these two docs as it is already in the method signature

File: src/Parser/GoogleSearchParser.php [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+            }
+
+            $results[] = new SearchResult(
+                $this->getUrl($xpath, $resultContainer),
+                $this->getTitle($xpath, $resultContainer),
+                $this->getDescription($xpath, $resultContainer),
+            );
+        }
+
+        return $results;
```
Comment (PeeHaa) [inline src/Parser/GoogleSearchParser.php]: Maybe this should just return the instance of the `SearchResults` directly instead of an array?

File: src/Plugin.php [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+{
+    private Client $httpClient;
+
+    public function __construct(Client $httpClient)
+    {
+        $this->httpClient = $httpClient;
+    }
+
+    /**
+     * @param string $term
```
Comment (PeeHaa) [inline src/Plugin.php]: Drop doc

File: src/ValueObject/Result/SearchResult.php [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+
+namespace AsyncBot\Plugin\GoogleSearch\ValueObject\Result;
+
+final class SearchResult
+{
+    private string $url;
+
+    private string $title;
+
+    private string $description;
```
Comment (PeeHaa) [inline src/ValueObject/Result/SearchResult.php]: nullable

File: src/Exception/UnexpectedHtmlFormat.php [hunk 7]
```diff
@@ -0,0 +1,11 @@
+<?php declare(strict_types=1);
+
+namespace AsyncBot\Plugin\GoogleSearch\Exception;
+
+class UnexpectedHtmlFormat extends Exception
+{
+    public function __construct(string $element)
+    {
+        parent::__construct(sprintf('Could not find the `%s` element in the document.', $element));
```
Comment (PeeHaa) [inline src/Exception/UnexpectedHtmlFormat.php]: Use double quotes instead of backticks as in the rest of the plugins

File: tests/Unit/ValueObject/SearchResultTest.php [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+    {
+        $this->assertSame('https://heap.space', $this->searchResult->getUrl());
+    }
+
+    public function testGetTitle(): void
+    {
+        $this->assertSame('HeapSpace', $this->searchResult->getTitle());
+    }
+
+    public function testGetDescription(): void
```
Comment (PeeHaa) [inline tests/Unit/ValueObject/SearchResultTest.php:27]: Add test for description when `null`

File: tests/Unit/Parser/ParseSearchResultsTest.php [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+
+    public function testParserThrowsWhenMissingMainContainer(): void
+    {
+        $this->expectException(UnexpectedHtmlFormat::class);
+        $this->expectExceptionMessage('Could not find the `main` element in the document.');
+
+        (new GoogleSearchParser())->parse(
+            $this->getDomFromFakeResponse('missing-main-container.html'),
+        );
+    }
```
Comment (PeeHaa) [inline tests/Unit/Parser/ParseSearchResultsTest.php:37]: - Add test when description container does not exist
- Add test when description is an empty string

File: src/Parser/GoogleSearchParser.php [hunk 10]
```diff
@@ -0,0 +1,115 @@
...
+        $linkNodes = $xpath->query('.//a', $resultContainer);
+
+        return $linkNodes->item(0);
+    }
+
+    private function getUrl(\DOMXPath $xpath, \DOMNode $resultContainer): string
+    {
+        $linkNode = $this->getLinkNode($xpath, $resultContainer);
+
+        return preg_replace('/^\/url\?q=/', '', $linkNode->getAttribute('href'));
```
Comment (PeeHaa) [inline src/Parser/GoogleSearchParser.php]: Use `~` as regex pattern delimiter so you don't have to escape slashes

**Review by PeeHaa — CHANGES_REQUESTED** [review]

File: tests/Unit/Parser/ParseSearchResultsTest.php (hunk 9 above)
Comment (ekinhbayar) [reply tests/Unit/Parser/ParseSearchResultsTest.php:37]: Can't test the first case as if the description container doesn't exist, it never gets added to the result set.

**Review by PeeHaa — APPROVED** [review]
